### ISSUE TITLE
Kanikos's Waypoint ODR image failing in K8s 1.23 (run with --force flag message)

### ISSUE THREAD

**Issue by dcanadillas** [hashicorp/waypoint#2984]
If using the On Demand Runner for a Docker build in Kubernetes 1.23 (it is not happening in 1.21) I receive a Kaniko error:

```
$ waypoint build -app back -local=false 

» Building back...
  Performing operation on "kubernetes" with runner profile "kubernetes"

» Cloning data from Git
  URL: https://gitlab.com/dcanadillas/bk8s-demo
  Ref: master
  Auth: username/password
✓ Building Docker image with kaniko...
✓ Testing registry and uploading entrypoint layer
❌ Executing kaniko...
 │ kaniko should only be run inside of a container, run with the --force flag if yo
 │ u are sure you want to continue
! exit status 1
```

**Steps to Reproduce**
* Install Waypoint from scratch with default values and connect to context:
```
helm install waypoint -n waypoint hashicorp/waypoint

waypoint login --from-kubernetes --from-kubernetes-namespace waypoint

```
* Initialize the project with a `waypoint.hcl` is something like this:
```
project = "back_front-example"

variable "registry" {
  default = "hcdcanadillas"
}
variable "platform" {
  default = "linux/amd64"
}
variable "back_version" {
  default = "v1"
}
variable "docker_creds" {
  type = object({
    username = string
    password = string
  })
}

runner {
  enabled = true
}

app "back" {
  path = "${path.project}/back"
  labels = {
      "service" = "back",
      "env" = "dev"
  }
  url {
    auto_hostname = false
  }

  build {
    use "docker" {
      dockerfile = "${path.app}/Dockerfile"
      buildkit = false
      platform = var.platform
    }
    registry {
      use "docker" {
        image = "${var.registry}/bk8s-back"
        tag   = "${var.back_version}-${regex("[^/]+$",var.platform)}"
        username = var.docker_creds.username
        password = var.docker_creds.password
      }
    }
  }

  deploy {
    use "kubernetes-apply" {
      path        = templatedir("${path.project}/k8s_templates/back-k8s")
      prune_label = "app=back"
    
  }
}
```

* Build the project (variables values already set):
  ```
  $ waypoint build -app back -local=false
  ...
  ...
  ✓ Building Docker image with kaniko...
  ✓ Testing registry and uploading entrypoint layer
  ❌ Executing kaniko...
   │ kaniko should only be run inside of a container, run with the --force flag if yo
   │ u are sure you want to continue
  ! exit status 1
  ```

**Expected behavior**
The project should build with Kaniko, as it is working with the same configuration in another cluster, but version 1.21

**Waypoint Platform Versions**

Here are the versions used:
```
$ waypoint version
CLI: v0.7.1 (062857d16)
Server: v0.7.1

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.3", GitCommit:"816c97ab8cff8a1c72eccca1026f7820e93e0d25", GitTreeState:"clean", BuildDate:"2022-01-25T21:25:17Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.1", GitCommit:"86ec240af8cbd1b60bcc4c03c20da9b98005b92e", GitTreeState:"clean", BuildDate:"2021-12-16T11:34:54Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
```

Plugin used is Docker for buikld

**Additional context**
I have tried the same exact configuration in a Kubernetes server 1.21 version cluster and it works good. All of them are Minikube clusters.

**Comment by briancain** [comment]
Thanks for the share @dcanadillas - I use Kubernetes version 1.22 locally and have no issues. I recommend using that version until we can figure out what's going on, thanks!

Also if you could, it would be good to see the full logs from running `waypoint build` with your ODR pod. Thank you!

**Comment by dcanadillas** [comment]
Thanks @briancain !  Yes, it is also working for me in v1.21. Didn't try on v1.22.  But I cannot make it work in v1.23.

I am attaching the complete logs of the task pod with the `odr` in the following file: 

[odr-pod.log](https://github.com/hashicorp/waypoint/files/8026504/odr-pod.log)

The error there is also referencing to the `--force` part of the kaniko build:

```
2022-02-08T18:22:22.989Z [DEBUG] waypoint.runner.agent.runner.app.back.builder.waypoint: executing kaniko: job_id=01FVD94T45ZMEVE888NJ8XVE3K job_op=*gen.Job_Build @module=waypoint.runner.agent.runner.app.back.build args=[/kaniko/executor, --context, dir:///kaniko/tmp/waypoint3732852373/back, -f, Dockerfile, -d, localhost:45853/hcdcanadillas/bk8s-back:v1-amd64] timestamp=2022-02-08T18:22:22.989Z
2022-02-08T18:22:23.016Z [WARN]  waypoint.runner.agent.runner.app.back.build: error during local operation: id=01FVD94XE7H0G9QZTXDQPR8VC7 job_id=01FVD94T45ZMEVE888NJ8XVE3K job_op=*gen.Job_Build err="rpc error: code = Unknown desc = exit status 1"
2022-02-08T18:22:23.018Z [DEBUG] waypoint.runner.agent.runner.app.back.build: metadata marked as complete: id=01FVD94XE7H0G9QZTXDQPR8VC7 job_id=01FVD94T45ZMEVE888NJ8XVE3K job_op=*gen.Job_Build
2022-02-08T18:22:23.018Z [DEBUG] waypoint.runner.agent.runner.app.back.registry.stdio: received EOF, stopping recv loop: job_id=01FVD94T45ZMEVE888NJ8XVE3K job_op=*gen.Job_Build err="rpc error: code = Unavailable desc = error reading from server: EOF"
```

I tested a standalone kaniko pod in the same 1.23 cluster, working as expected:

```
export REGISTRY_SERVER="https://index.docker.io/v1/"
export REGISTRY_USER=dcanadillas
export REGISTRY_PASS=<my_api_secret>
export REGISTRY_EMAIL=<my_docker_email>

kubectl create secret \
docker-registry dockerhub-registry \
--docker-server=$REGISTRY_SERVER \
--docker-username=$REGISTRY_USER \
--docker-password=$REGISTRY_PASS \
--docker-email=$REGISTRY_EMAIL

kubectl apply -f - <<EOF
---
apiVersion: v1
kind: Pod
metadata:
  name: kaniko-builder
spec:
  containers:
    - name: kaniko
      image: gcr.io/kaniko-project/executor:latest
      args: ["--context=git://github.com//Imoustak/kaniko-build-demo.git",
             "--destination=dcanadillas/kaniko-build-demo:1.0.0",
             "--dockerfile=dockerfile"]
      volumeMounts:
        - name: kaniko-secret
          mountPath: /kaniko/.docker
  restartPolicy: Never
  volumes:
    - name: kaniko-secret
      secret:
        secretName: dockerhub-registry
        items:
          - key: .dockerconfigjson
            path: config.json
EOF
```

And the logs of that standalone Kaniko build was:

```
Enumerating objects: 13, done.
Counting objects: 100% (13/13), done.
Compressing objects: 100% (10/10), done.
Total 13 (delta 2), reused 6 (delta 1), pack-reused 0
INFO[0001] Retrieving image manifest ubuntu             
INFO[0001] Retrieving image ubuntu from registry index.docker.io 
INFO[0003] Built cross stage deps: map[]                
INFO[0003] Retrieving image manifest ubuntu             
INFO[0003] Returning cached image manifest              
INFO[0003] Executing 0 build triggers                   
INFO[0003] Skipping unpacking as no commands require it. 
INFO[0003] ENTRYPOINT ["/bin/bash", "-c", "echo hello"] 
INFO[0003] Pushing image to dcanadillas/kaniko-build-demo:1.0.0 
INFO[0005] Pushed image to 1 destinations
```

So it seems that the kaniko image works outside, but not in the ODR container.

 I found a similar issue here: https://github.com/GoogleContainerTools/kaniko/issues/1542

Hope this helps.

**Comment by briancain** [comment]
Interesting, so potentially an upstream issue with Kaniko figuring out if it's inside a container? Anyway thanks for all the extra info!

**Comment by dcanadillas** [comment]
I have found a workaround just by setting `container` environment variable in the runner to `docker` value (thanks to this other issue comment: https://github.com/GoogleContainerTools/kaniko/issues/1542#issuecomment-853929795).

So, if I set a runner profile like this:

```
waypoint runner profile set -name kubernetes \
-plugin-type kubernetes -\
-env-vars "container=docker" \
-default
```

It just works with that environment variable definition:

```
$ waypoint build -app back

» Building back...
  Performing operation on "kubernetes" with runner profile "kubernetes"

» Cloning data from Git
  URL: git@gitlab.com:dcanadillas/bk8s-demo.git
  Ref: master
  Auth: ssh
✓ Building Docker image with kaniko...
✓ Testing registry and uploading entrypoint layer
✓ Executing kaniko...
 │ flicting behaviour with the system package manager. It is recommended to use a v
 │ irtual environment instead: https://pip.pypa.io/warnings/venv
 │ WARNING: You are using pip version 21.2.4; however, version 22.0.3 is available.
 │ You should consider upgrading via the '/usr/local/bin/python -m pip install --up
 │ grade pip' command.
 │ INFO[0006] Taking snapshot of full filesystem...
 │ INFO[0006] CMD exec gunicorn --bind :$PORT --workers 1 --threads 8 --timeout 0 m
 │ ain:app
 │ INFO[0006] Pushing image to localhost:34825/dcanadillas/bk8s-back:v1-amd64
 │ INFO[0015] Pushed image to 1 destinations
✓ Image pushed to 'dcanadillas/bk8s-back:v1-amd64'
```

I don't know what is the difference between v1.22 and v1.23 or if it is another versioning issue, but the environment variable workaround makes the trick.

**Comment by xiaolin-ninja** [comment]
Thanks for bringing this to our attention, we'll be sure to patch this bug in Kaniko.

**Comment by rastakajakwanna** [comment]
Kaniko is failing also on EKS 1.21 and the referred bug report on Google Tools helped me to sort it out. My path started here and that's why I wanted to share it here in case someone else would arrive here with the same problem.

It was a sudden change from Kaniko working to Kaniko failing with error "cannot push layer: 500 Internal server error".

First, I tried setting container=docker. Just in case you would also just copy paste like I did - don't, check the runner IDs and apply to the default runner, the above would create new config which is never applied unless you do further steps. Anyway, it did not help.

Second, I tried setting container=kube as suggested by Kaniko developers somewhere deeper in the rabbit hole. It did not help.

For the Xth time, I tried to change container version as I've noticed that the default waypoint setting points to hashicorp Docker registry and tag latest, while Helm version points to specific Docker Hub image tag.

Finally, changing image from "docker.io/hashicorp/waypoint-odr:0.8.1" to "docker.io/hashicorp/waypoint-odr:0.8.0" solved my issue with Kaniko being unable to push layer. The final command which resolved my problem is
```
waypoint runner profile set -name <ID-OF-YOUR-RUNNER-IN-K8S> -plugin-type kubernetes -env-var "container=docker" -default -oci-url=docker.io/hashicorp/waypoint-odr:0.8.0
```
Honestly, I did not test variant without the env-var (env-vars from the command above is deprecated), because I did not find a way how to remove this part of the configuration (is there a delete api?) and not specifying it did not remove it and specifying empty value resulted in error message.

**Comment by briancain** [comment]
Hey all - I've opened a [pull request](https://github.com/hashicorp/waypoint/pull/3322) to automatically inject `container=docker` for the Kubernetes task plugin in the on-demand runners environment so that Kaniko is properly picked up and used. We'll use this as a temp fix for now until the upstream Kaniko fix has been released. Thanks again for all of the information provided in this thread! We appreciate your investigation into this issue ❤️
